lib/game.rb

```diff
@@ -0,0 +1,93 @@
+class Game
+
+    attr_reader :score, :frame_counter
+    FRAME_MAX = 10
+    BONUS_VALUE = 10
+
+    def initialize
+        @score = 0
+        @frame_counter = 0
+        @spare_counter = 0
+        @extra_roll = 0
+        @results = []
+        @frame = []
+    end
+
+    def roll(pins)
+        fail 'This game is over' if @frame_counter == FRAME_MAX
+        @results << pins
+        @frame  << pins
+        bonus_determiner
```

Comment:
Looks like so far you have one class `Game` which calculates everything - maybe you could start to think on how to extract some concerns to other class, and have `Game` delegate some work to these new classes. For example, everything related to bonus calculation could go into its own class, the code of `roll` could look like this:

```irb
def roll
   fail 'This game is over' if @frame_counter == FRAME_MAX
  @results << pins
  @frame  << pins
  
  @score += @bonus_calculator.calculate_added_bonus(pins)

  # etc
end
```

And following the same process, you can extract maybe the part that calculates the regular frame score, etc. Note that this is one way of splitting this class, but you could find other ways, and this is fine too!